### PR TITLE
Simplify `ErrBalanceTx`.

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue

ADP-3124

## Description

This PR simplifies the `ErrBalanceTx` type.

## Motivation

1. Require less nesting when pattern matching on values of `ErrBalanceTx` that originate from `cardano-coin-selection`.
2. Define all errors for `balanceTransaction` in one place.
3. Eliminate usages of the `WalletSelectionContext` type in the public type signature of `balanceTransaction`.
4. Eliminate imports of internal coin selection modules within `cardano-wallet`.
5. Make it easier for future PRs to change the types of error fields to ledger-compatible types.

### CODE REVIEW

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 1]
```diff
@@ -30,11 +30,6 @@ import Cardano.Address.Script
     ( Cosigner (..) )
 import Cardano.Ledger.Alonzo.TxInfo
     ( TranslationError (..) )
-import Cardano.Tx.Balance.Internal.CoinSelection
-    ( SelectionBalanceError (..)
-    , UnableToConstructChangeError (..)
-    , WalletSelectionContext
-    )
```
Comment (jonathanknowles) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs:37]: We no longer rely on internal coin selection modules in the wallet HTTP error API. 🍾

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 2]
```diff
@@ -3226,25 +3221,22 @@ prop_balanceTransactionValid
...
-                            property True
-                        (False, True) ->
-                            label "missing coin" $
-                            property True
-                        (True, False) ->
-                            label "missing tokens" $
-                            counterexample (show err) $ property True
-                        (True, True) ->
-                            property False
+            Left (ErrBalanceTxAssetsInsufficient err) -> do
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:3224]: Much less nesting than before. 🍾

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs [hunk 3]
```diff
@@ -298,6 +333,12 @@ data ErrBalanceTx
     | ErrBalanceTxInputResolutionConflicts (NonEmpty (W.TxOut, W.TxOut))
     | ErrBalanceTxUnresolvedInputs (NonEmpty W.TxIn)
     | ErrBalanceTxOutputError ErrBalanceTxOutputError
+    | ErrBalanceTxUnableToCreateChange ErrBalanceTxUnableToCreateChangeError
+    | ErrBalanceTxUnableToCreateInput
+    -- ^ Returned when __both__ of the following conditions are true:
+    --   - the given partial transaction has no existing inputs; and
+    --   - the given UTxO index is empty.
+    -- A transaction must have at least one input in order to be valid.
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs:341]: 👍

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 4]
```diff
@@ -553,6 +551,22 @@ instance IsServerError ErrBalanceTx where
                 , fmt $ blockListF' "-" conflictF conflicts
                 ]
         ErrBalanceTxOutputError err -> toServerError err
+        ErrBalanceTxUnableToCreateInput ->
```
Comment (Anviking) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs:554]: On second thought, I'm not sure about the notion of "creating inputs". We're not really "creating" inputs, we're just selecting or adding them.

Would `UnableToSelectInput` be better? Or perhaps `EmptyUTxO` wasn't that bad after all, even if there are cases where we _can_ balance with an empty utxo set? 🤔

**Review by Anviking — DISMISSED** [review]
thanks — lgtm! One concern about naming - maybe we can discuss tomorrow?

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs (hunk 4 above)
Comment (jonathanknowles) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs:554]: > On second thought, I'm not sure about the notion of "creating inputs". We're not really "creating" inputs, we're just selecting or adding them.

The way I rationalised it to myself was that:
- we _select_ existing UTxOs from the set of available UTxOs
- we _create_ a transaction input that _references_ a UTxO 
 
Note that this is just my own mental model.

> Would `UnableToSelectInput` be better? Or perhaps `EmptyUTxO` wasn't that bad after all, even if there are cases where we _can_ balance with an empty utxo set? 🤔

Out of these two, I'd be in favour of `UnableToSelectInput`.

But would be happy to discuss with you tomorrow! If we come to a new consensus, we can create a PR that updates the name.

**Comment by jonathanknowles** [comment]
> maybe we can discuss tomorrow?

Yes, happy to discuss tomorrow!

I'll merge the existing changes now -- but we can make another PR that tweaks the names if we come to a new consensus.

**Review by Anviking — APPROVED** [review]

**Pull request merged**